sbi/diagnostics/sbc.py

```diff
@@ -0,0 +1,221 @@
+                    )
+                )
+                pbar.update(sbc_batch_size)
+
+    # Aggregate results.
+    ranks = []
+    dap_samples = []
+    log_probs = []
+    for out in sbc_outputs:
+        ranks.append(out[0])
+        log_probs.append(out[1])
+        dap_samples.append(out[2])
+
+    ranks = torch.cat(ranks)
+    dap_samples = torch.cat(dap_samples)
+    log_probs = torch.cat(log_probs)
+
+    return ranks, log_probs, dap_samples
+
+
+def sbc_on_batch(thos, xos, posterior, L, ranking_rv):
+    """Return SBC results for a batch of SBC parameters and data from prior.
+
+    Returns
+        ranks: ranks of true parameters vs. posterior samples under the specified RV,
+            for each posterior dimension.
+        log_prob_thos: log prob of true parameters under the approximate posterior.
+        dap_samples: samples from the data averaged posterior for the current batch,
+            i.e., a single sample from each approximate posterior.
+    """
+
+    log_prob_thos = torch.zeros(thos.shape[0])
+    dap_samples = torch.zeros_like(thos)
+    ranks = torch.zeros_like(thos)
+
+    for idx, (tho, xo) in enumerate(zip(thos, xos)):
+        # Log prob of true params under posterior.
+        log_prob_thos[idx] = posterior.log_prob(tho, x=xo)
+
+        # Rank true params vs posterior params under ranking_rv.
+        ths = posterior.sample((L,), x=xo, show_progress_bars=False)
+        dap_samples[idx] = ths[
+            0,
+        ]
+        # rank for each posterior dimension.
+        for dim in range(thos.shape[1]):
+            ranks[idx, dim] = (
+                (ranking_rv.log_prob(ths[:, dim]) < ranking_rv.log_prob(tho[dim]))
+                .sum()
+                .item()
+            )
+
+    return ranks, log_prob_thos, dap_samples
+
+
+def sbc_checks(
+    ranks: Tensor,
+    log_probs: Tensor,
+    prior_samples: Tensor,
+    dap_samples: Tensor,
+    num_ranks: int,
+):
+    """Return uniformity checks, data averaged posterior checks and NLTP for SBC."""
+    if ranks.shape[0] < 100:
+        warnings.warn(
+            """You are computing SBC checks with less than 100 samples. These checks
+            should be based on a large number of test samples theta_o, x_o. We
+            recommend using at least 100."""
+        )
+
+    ks_pvals, c2st_ranks = check_uniformity(ranks, num_ranks)
+    c2st_scrores_dap = check_prior_vs_dap(prior_samples, dap_samples)
+    nltp = torch.mean(-log_probs)
+
+    return dict(
+        ks_pvals=ks_pvals,
+        c2st_ranks=c2st_ranks,
+        c2st_dap=c2st_scrores_dap,
+        nltp=nltp,
+    )
+
+
+def check_prior_vs_dap(prior_samples: Tensor, dap_samples: Tensor):
+    """Return the c2st accuracy between prior and data avaraged posterior samples.
+
+    c2st is calculated for each dimension separately.
+    """
+
+    assert prior_samples.shape == dap_samples.shape
+
+    return torch.tensor(
+        [
+            c2st(s1.unsqueeze(1), s2.unsqueeze(1))
```

Comment:
Quite a cannon that is shot at this problem - I understand the motivation though. I naively wonder if a kNN classifier would suffice as well and be a bit quicker. :man_shrugging:

Reply:
I undertook some experiments, if `c2st` should be used here, I suggest to choose a more lightweight classifier, see my proposal for a randomforest in #503 

the more dimensions, prior_samples exposes, the more costly will the MLP in `c2st` slow things down.